data-serving/data-service/src/controllers/case.ts

```diff
@@ -123,6 +123,42 @@ export const update = async (req: Request, res: Response): Promise<void> => {
     }
 };
 
+/**
+ * Upserts a case based on a compound index of
+ * caseReference.{dataSourceId, dataEntryId}.
+ *
+ * On success, the returned status code indicates whether than item was created
+ * (201) or updated (200).
+ *
+ * Handles HTTP PUT /api/cases.
+ */
+export const upsert = async (req: Request, res: Response): Promise<void> => {
+    try {
+        const c = await Case.findOne({
```

Comment:
you'll want a transaction to do the fetch + save here now?

https://docs.mongodb.com/manual/core/transactions/#transactions-api

Reply:
Totally, this isn't atomic. A transaction would be better, but for now we're at least protected from doing anything "unsafe" by [Mongoose versioning](https://mongoosejs.com/docs/guide.html#versionKey). It would be better to do this in a transaction, but given the expected QPS (to a given document), I figured it was OK to revisit.